src/CoreFn/FromJSON.purs

```diff
@@ -0,0 +1,377 @@
+module CoreFn.FromJSON
+  ( moduleFromJSON
+  ) where
+
+import Prelude
+
+import Control.Alt ((<|>))
+import CoreFn.Ann (Ann(..), Comment(..), SourcePos(..), SourceSpan(..))
+import CoreFn.Binders (Binder(..))
+import CoreFn.Expr (Bind(..), CaseAlternative(..), Expr(..))
+import CoreFn.Ident (Ident(..))
+import CoreFn.Literal (Literal(..))
+import CoreFn.Meta (ConstructorType(..), Meta(..))
+import CoreFn.Module (FilePath(..), Module(..), ModuleImport(..), Version(..))
+import CoreFn.Names (ModuleName(..), ProperName(..), Qualified(..))
+import Data.Array as Array
+import Data.Either (Either(..))
+import Data.Foreign (F, Foreign, ForeignError(..), fail, readArray, readBoolean, readChar, readInt, readNull, readNumber, readString, typeOf)
+import Data.Foreign.Index (index, readProp)
+import Data.Foreign.JSON (parseJSON)
+import Data.Foreign.Keys (keys)
+import Data.Maybe (Maybe(..))
+import Data.Newtype (unwrap)
+import Data.Traversable (traverse)
+import Data.Tuple (Tuple(..), uncurry)
+
+objectType :: String
+objectType = "object"
+
+object :: forall a. (Foreign -> F a) -> Foreign -> F a
+object _ json
+  | typ <- typeOf json, typ /= objectType = fail $ TypeMismatch objectType typ
+object f json = f json
+
+constructorTypeFromJSON :: Foreign -> F ConstructorType
+constructorTypeFromJSON json = do
+  type_ <- readString json
+  case type_ of
+    "ProductType" -> pure ProductType
+    "SumType" -> pure SumType
+    _ -> fail $ ForeignError $ "Unknown ConstructorType: " <> type_
+
+metaFromJSON :: Foreign -> F Meta
+metaFromJSON = object $ \json -> do
+  type_ <- readProp "metaType" json >>= readString
+  case type_ of
+    "IsConstructor" -> isConstructorFromJSON json
+    "IsNewType" -> pure IsNewtype
```

Comment:
The test failing here might be because of the casing. Probably needs to be `"IsNewtype"`.